File: ex/uri1211.cpp
#include <iostream>
#include <stdlib.h>
#include <map>
using namespace std;

int walk(string phone, map<char, void*> *tree, int index)
{
	if (index == phone.length())
		return 0;

	char digit = phone.at(index);
	int result = 0;
	void *ptr;

	if (tree->count(digit) > 0) {
		ptr = (*tree)[digit];
		result = walk(phone, (map<char, void*>*) ptr, index+1) + 1;
	}
	else {
		(*tree)[digit] = new map<char, void*>();
		ptr = (*tree)[digit];
		result = walk(phone, (map<char, void*>*) ptr, index+1);
	}

	return result;
}

int main(int argc, char const *argv[]) {
	int n, economy;
	string phone;
	void *phones;

	while (cin >> n)
	{
		economy = 0;
		phones = new map<char, void*>();

		for (size_t i = 0; i < n; i++) {
			cin >> phone;
			economy += walk(phone, (map<char, void*>*) phones, 0);
		}

		cout << economy << endl;
		free(phones);
	}

	return 0;
}

File: src/uri/uri1211.py
def walk(phone, tree, index):
    if index == len(phone):
        return 0

    digit = phone[index]
    if tree.has_key(digit):
        return walk(phone, tree[digit], index+1) + 1
    else:
        tree[digit] = {}
        return walk(phone, tree[digit], index+1)

def run(phone, book):
    tree = book
    economy = 0

    for digit in phone:
        if tree.has_key(digit):
            economy += 1
        else:
            tree[digit] = {}
        tree = tree[digit]

    return economy

def setup():
    global phones

    phones = []
    for x in xrange(int(raw_input())):
        phones.append(raw_input())

def draw():
    global phones

    book = {}
    economy = 0

    for phone in phones:
        # economy += walk(phone, book, 0)
        economy += run(phone, book)

    print(economy)

if __name__ == '__main__':
    while True:
        try:
            setup()
            draw()
        except:
            break
